File: 547-number-of-provinces/number-of-provinces.cpp
class Solution {
public:
    void dfs(int at,  vector<vector<int>> aL, vector<bool> &vis){
        vis[at] = true;
        for(int j = 0; j < aL.size(); j++){
            if(aL[at][j] == 1 && !vis[j]){
                dfs(j, aL, vis);
            }
        }
    }
    int findCircleNum(vector<vector<int>>& isConnected) {
        //TC - O(N + V + 2E) and SC - O( N + N) for vis arrays and recursion stack space 
        int n = isConnected.size();
        vector<bool> vis(n, false);
        //Conversion of Adjacency matrix to List
        /*
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                if(isConnected[i][j] == 1 && i != j){
                    adjList[i].push_back(j);
                }
            }
        }
        */

        //Simple algorithm to count the number of different DFS calls made thart will the num of provinces
        int ans = 0;
        for(int i = 0; i< n; i++){
            if(!vis[i]){
                ans++;
                dfs(i, isConnected, vis);
            }
        }
        return ans;
    }
};

File: 547-number-of-provinces/number-of-provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        adj = defaultdict(list)
        for u in range(n):
            for v in range(n):
                if isConnected[u][v] == 1:
                    adj[u].append(v)
                    adj[v].append(u)

        vis = [0]*(n+1)
        def dfs(node):
            if vis[node] == 1:
                return
            vis[node] = 1
            for neigh in adj[node]:
                dfs(neigh)

        ans = 0
        for i in range(n):
            if vis[i] == 0:
                dfs(i)
                ans += 1
        return ans